control-plane-operator/controllers/hostedcontrolplane/hostedcontrolplane_controller.go

```diff
@@ -407,21 +415,133 @@ func (r *HostedControlPlaneReconciler) ensureInfrastructure(ctx context.Context,
 	if err := ensureVPNSCC(r, hcp, targetNamespace); err != nil {
 		return status, fmt.Errorf("failed to ensure privileged SCC for the new namespace: %w", err)
 	}
+	if hcp.Spec.Services == nil {
+		return status, fmt.Errorf("service publishing strategy undefined")
+	}
+	for _, serviceItr := range hcp.Spec.Services {
```

Comment:
You could extract a function for dealing with each publishing type which could be reused by all these cases

Reply:
I'm not sure how much code reuse we want to dictate across each type considering that really in each path it's create/reconcile a service and then fetch unique pieces of data depending on the type of service. Maybe I'm missing what you are suggesting though.